kore/src/Kore/Logger.hs

```diff
@@ -240,18 +295,82 @@ withLogScope newScope = localLogAction (contramap appendScope)
 -- ---------------------------------------------------------------------
 -- * LoggerT
 
+class (Typeable entry, Pretty.Pretty entry) => Entry entry where
+    toEntry :: entry -> SomeEntry
+    toEntry = SomeEntry
+
+    fromEntry :: SomeEntry -> Maybe entry
+    fromEntry (SomeEntry entry) = Data.Typeable.cast entry
+
+    shouldLog :: Severity -> Set Scope -> entry -> Bool
+
+data SomeEntry where
+    SomeEntry :: Entry entry => entry -> SomeEntry
+
+instance Pretty.Pretty SomeEntry where
+    pretty (SomeEntry entry) = Pretty.pretty entry
+
+withSomeEntry
+    :: (forall entry. Entry entry => entry -> a)
+    -> SomeEntry
+    -> a
+withSomeEntry f (SomeEntry entry) = f entry
+
+instance Entry LogMessage where
+    shouldLog :: Severity -> Set Scope -> LogMessage -> Bool
+    shouldLog minSeverity currentScope LogMessage { severity, scope } =
+        severity >= minSeverity && scope `member` currentScope
+      where
+      member :: [Scope] -> Set Scope -> Bool
+      member s set
+        | Set.null set = True
+        | otherwise    = Fold.any (`Set.member` set) s
+
+class Monad m => MonadLog m where
+    logM :: Entry entry => entry -> m ()
+    localScope :: Entry e1 => Entry e2 => (e1 -> e2) -> m a -> m a
+
 newtype LoggerT m a =
-    LoggerT { getLoggerT :: ReaderT (LogAction m LogMessage) m a }
+    LoggerT { getLoggerT :: ReaderT (LogAction m SomeEntry) m a }
     deriving (Functor, Applicative, Monad)
     deriving (MonadIO)
 
-instance MonadTrans LoggerT where
-    lift = LoggerT . Monad.Trans.lift
-    {-# INLINE lift #-}
+instance Monad m => MonadLog (LoggerT m) where
+    logM entry =
+        LoggerT $ ask >>= Monad.Trans.lift . (<& toEntry entry)
+    localScope f (LoggerT (ReaderT logActionReader)) =
+        LoggerT . ReaderT
+            $ \(LogAction logAction) ->
+                logActionReader . LogAction
+                    $ \entry ->
+                        case fromEntry entry of
+                            Nothing -> logAction entry
+                            Just entry' -> logAction $ toEntry $ f entry'
 
 instance Monad m => WithLog LogMessage (LoggerT m) where
-    askLogAction = LoggerT . fmap liftLogAction $ ask
+    askLogAction :: LoggerT m (LogAction (LoggerT m) LogMessage)
+    askLogAction = LoggerT . ReaderT $ pure . go
```

Comment:
This won't work since we can't go from `LogAction m msg -> LogAction m msg` to `msg -> msg` (needed in `logScope`). I think we can do it the other way around: provide default implementations of `MonadLog` for anything with a `WithLog` instance. Would that do us any good?